Miniview Redesign

Bjvanminnen opened the pull request:
Without redesign flag:
![image](https://cloud.githubusercontent.com/assets/1767466/24530702/9435fb2e-1568-11e7-9e4e-adef6ddde20f.png)

With flag, summary view
![image](https://cloud.githubusercontent.com/assets/1767466/24530710/a0477316-1568-11e7-8ade-9f799db770cf.png)

With flag, detail view
![image](https://cloud.githubusercontent.com/assets/1767466/24530713/aabd55a4-1568-11e7-929f-4dffb6967080.png)

With flag, grouped 
![image](https://cloud.githubusercontent.com/assets/1767466/24620294/f0fa6d72-1851-11e7-89c4-ea434952233a.png)

Bjvanminnen commented on apps/src/code-studio/header.js:
```diff
@@ -108,10 +112,6 @@ header.build = function (scriptData, stageData, progressData, currentLevelId, pu
   });
-  $('.header_popup').click(function (e) {
-    e.stopPropagation(); // Clicks inside the popup shouldn't close it
-  });
```
This was making it so that all clicks inside of the popup were ignored, and I couldn't use the toggle. Instead I moved this logic into hideHeaderPopup.

Bjvanminnen replied:
In an ideal world, a lot of the logic here gets moved into a React component. I ultimately decided that was out of scope for now.

davidsbailey commented:
This is my first real look at this feature so I have a few initial questions:
1. is there a spec?
2. what is the difference between "with flag, summary view" and "with flag, grouped"? Screenshots of the same stage in both views might be helpful.
3. one difference between "summary view" and "grouped" is the purple "Content" frame. What is the purpose of this frame?

Bjvanminnen commented:
1) Spec is here https://docs.google.com/document/d/1dpJrFKi8G5wluPBbk0kxjYcOOAiKTE2zC6iG4a_LbSY/edit (but covers a lot more than just the miniview).
2) A particular script will either have grouped lessons or not
3) This is the idea behind scripts with grouping. Stages are assigned categories (Content in this case) and all stages with the same category end up in the same category. Inside of a group, you can still toggle between summary/detail views of the tables. For ungrouped scripts, there will be a single table and no purple.

It's perhaps worth noting that the grouping stuff is not new to this PR (it also shows up on the s/<script_name> pages for grouped scripts). This PR just makes sure that it and other views still look okay in the miniview aka progress dropdown.

davidsbailey commented:
I am seeing this warning message at http://localhost-studio.code.org:3000/s/course3/stage/2/puzzle/1?enableExperiments=progressRedesign when the window has been open for a while:
<img width="541" alt="screen shot 2017-04-04 at 1 15 28 pm" src="https://cloud.githubusercontent.com/assets/8001765/24676690/db02eb7e-1938-11e7-8a40-f2177394d886.png">

Bjvanminnen commented:
Were you doing anything when this happened, or did you just leave it running?

Do you have a full call stack?

davidsbailey commented:
I was able to catch it in the debugger. I was moving the mouse around a bunch over the non-grouped view, which causes a lot of tooltips to appear and then disappear. Here is the call stack:
```
printWarning	@	warning.js?8a56*****:36
warning	@	warning.js?8a56*****:61
getInternalInstanceReadyForUpdate	@	ReactUpdateQueue.js?fd2c*****:50
enqueueSetState	@	ReactUpdateQueue.js?fd2c*****:202
ReactComponent.setState	@	ReactComponent.js?702a*****:65
showTooltip	@	index.js?4b90:317
```
Which appears to be at this line: https://github.com/wwayne/react-tooltip/blob/master/standalone/react-tooltip.js#L734

Bjvanminnen commented:
Thanks. I suspect this error is not specific to this particular PR, but i will look into it further.

davidsbailey commented on apps/src/templates/progress/ProgressDetailToggle.jsx:
```diff
@@ -56,21 +65,27 @@ const ProgressDetailToggle = React.createClass({
   },
```
`setIsSummaryView` could be a good candidate for outputting to the action console via `storybook.action` in ProgressDetailToggle.story.jsx, however I'm not sure how to do that when redux is involved as it is in that story.

davidsbailey commented on apps/src/code-studio/components/progress/MiniView.jsx:
```diff
@@ -1,33 +1,82 @@
+  groupView: {
+    margin: 20
+  }
+};
```
possibly worth mentioning here the case with no margins

Bjvanminnen commented:
"It should say [View Unit Overview] even when you’re a teacher viewing a student’s progress (instead of saying “View ’s progress” like it does right now)."

Yep, this has been tested :)

Bjvanminnen commented:
Managed to repro the error you saw. Repro appears to be:
1) Open miniview
2) switch to detail view
3) mouseover bubble in miniview
4) mouseover bubble in header.

I don't repro in prod (where you can't toggle into detail view), so I'll fix this before merging.

Bjvanminnen commented:
@davidsbailey Made some small to medium changes (removed footer, changed link to a button, and fixed tooltip errors) if you'd like to take another quick look. Thanks for the feedback!

davidsbailey commented on apps/src/code-studio/components/progress/ProgressDot.jsx:
```diff
@@ -283,6 +279,8 @@ export const ProgressDot = Radium(React.createClass({
     const isLocked = status === LevelStatus.locked;
 
+    const tooltipId = _.uniqueId();
+
```
Nice find on the tooltip error!

It seems like this approach could theoretically lead to worse performance, since react will never see the same ID even when the same tooltip is being shown a second time. If the problem is that`props.level.title` is sometimes absent, then perhaps you could use a unique ID only in when it is absent. However if the problem is non-unique values of `props.level.title` then that probably won't work.

I think it's also fine to do nothing, and just keep this in mind if we start observing bad performance moving the mouse in an area of the screen that causes many tooltips to show.
